tests/componenttests/server/tests/mediaPipeline/EncryptedPlaybackTest.cpp

```diff
@@ -0,0 +1,342 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ActionTraits.h"
+#include "ConfigureAction.h"
+#include "ExpectMessage.h"
+#include "IMediaFrameWriter.h"
+#include "Matchers.h"
+#include "MediaPipelineTest.h"
+#include "MessageBuilders.h"
+#include "SegmentBuilder.h"
+
+namespace
+{
+constexpr unsigned kFramesToPush{1};
+constexpr int kFrameCountInPausedState{3};
+constexpr int kFrameCountInPlayingState{24};
+} // namespace
+
+using testing::_;
+using testing::Return;
+
+namespace firebolt::rialto::server::ct
+{
+class EncryptedPlaybackTest : public MediaPipelineTest
+{
+public:
+    EncryptedPlaybackTest() = default;
+    ~EncryptedPlaybackTest() override = default;
+
+    void willAddProtectionMetadata(const std::unique_ptr<IMediaPipeline::MediaSegment> &segment, GstBuffer &dataBuffer,
+                                   GstBuffer &keyIdBuffer, GstBuffer &initVectorBuffer, GstBuffer &subsamplesBuffer,
+                                   guint8 &subsamples, GstMeta &meta)
+    {
+        const std::size_t kSubsamplesSize{segment->getSubSamples().size() * (sizeof(guint16) + sizeof(guint32))};
+        EXPECT_CALL(*m_gstWrapperMock, gstBufferNewAllocate(nullptr, segment->getKeyId().size(), nullptr))
+            .WillOnce(Return(&keyIdBuffer))
+            .RetiresOnSaturation();
+        EXPECT_CALL(*m_gstWrapperMock, gstBufferFill(&keyIdBuffer, 0, _, segment->getKeyId().size()))
+            .WillOnce(Return(segment->getKeyId().size()))
+            .RetiresOnSaturation();
+        EXPECT_CALL(*m_gstWrapperMock, gstBufferNewAllocate(nullptr, segment->getInitVector().size(), nullptr))
+            .WillOnce(Return(&initVectorBuffer))
+            .RetiresOnSaturation();
+        EXPECT_CALL(*m_gstWrapperMock, gstBufferFill(&initVectorBuffer, 0, _, segment->getInitVector().size()))
+            .WillOnce(Return(segment->getInitVector().size()))
+            .RetiresOnSaturation();
+        EXPECT_CALL(*m_glibWrapperMock, gMalloc(kSubsamplesSize)).WillOnce(Return(&subsamples)).RetiresOnSaturation();
+        EXPECT_CALL(*m_gstWrapperMock, gstByteWriterInitWithData(_, &subsamples, kSubsamplesSize, FALSE));
+        EXPECT_CALL(*m_gstWrapperMock, gstByteWriterPutUint16Be(_, segment->getSubSamples().front().numClearBytes));
+        EXPECT_CALL(*m_gstWrapperMock, gstByteWriterPutUint32Be(_, segment->getSubSamples().front().numEncryptedBytes));
+        EXPECT_CALL(*m_gstWrapperMock, gstBufferNewWrapped(&subsamples, kSubsamplesSize)).WillOnce(Return(&subsamplesBuffer));
+        EXPECT_CALL(*m_gstWrapperMock, gstBufferAddMeta(&dataBuffer, _, _)).WillOnce(Return(&meta));
+    }
+
+    void pushEncryptedAudioData(int needDataFrameCount)
+    {
+        // First, generate new data
+        auto segment{SegmentBuilder().basicAudioSegment(m_audioSourceId).withEncryptionData()()};
```

Comment:
Do you think the encrypted data written to segments should be associated with a mediaKeySession instead of been random data? Not something to do now but maybe an improvement in the future to make this test more realistic.

Reply:
We don't have MediaKeys tests yet, but we can think about it. But it may be hard to combine MediaPipeline and MediaKeys tests in Rialto Server, because Gstreamer is wrapped and GstRialtoDecryptor element is not created